sycl/source/detail/graph_impl.cpp

```diff
@@ -1095,6 +1154,246 @@ void exec_graph_impl::duplicateNodes() {
+        this, Nodes, AllocaQueue, MRequirements, MExecutionEvents);
+  } else {
+    for (auto &Node : Nodes) {
+      updateImpl(Node);
+    }
+  }
+}
+
+void exec_graph_impl::updateImpl(std::shared_ptr<node_impl> Node) {
+  auto ContextImpl = sycl::detail::getSyclObjImpl(MContext);
+  const sycl::detail::PluginPtr &Plugin = ContextImpl->getPlugin();
+  auto DeviceImpl = sycl::detail::getSyclObjImpl(MGraphImpl->getDevice());
+
+  // Gather arg information from Node
+  auto &ExecCG =
+      *(static_cast<sycl::detail::CGExecKernel *>(Node->MCommandGroup.get()));
+  // Copy args because we may modify them
+  std::vector<sycl::detail::ArgDesc> NodeArgs = ExecCG.getArguments();
+  // Copy NDR desc since we need to modify it
+  auto NDRDesc = ExecCG.MNDRDesc;
+
+  pi_kernel PiKernel = nullptr;
+  auto Kernel = ExecCG.MSyclKernel;
+  auto KernelBundleImplPtr = ExecCG.MKernelBundle;
+  std::shared_ptr<sycl::detail::kernel_impl> SyclKernelImpl = nullptr;
+  const sycl::detail::KernelArgMask *EliminatedArgMask = nullptr;
+
+  // Use kernel_bundle if available unless it is interop.
+  // Interop bundles can't be used in the first branch, because the kernels
+  // in interop kernel bundles (if any) do not have kernel_id
+  // and can therefore not be looked up, but since they are self-contained
+  // they can simply be launched directly.
+  if (KernelBundleImplPtr && !KernelBundleImplPtr->isInterop()) {
+    auto KernelName = ExecCG.MKernelName;
+    kernel_id KernelID =
+        sycl::detail::ProgramManager::getInstance().getSYCLKernelID(KernelName);
+    kernel SyclKernel =
+        KernelBundleImplPtr->get_kernel(KernelID, KernelBundleImplPtr);
+    SyclKernelImpl = sycl::detail::getSyclObjImpl(SyclKernel);
+    PiKernel = SyclKernelImpl->getHandleRef();
+    EliminatedArgMask = SyclKernelImpl->getKernelArgMask();
+  } else if (Kernel != nullptr) {
+    PiKernel = Kernel->getHandleRef();
+    auto SyclProg = Kernel->getProgramImpl();
+    EliminatedArgMask = Kernel->getKernelArgMask();
+  } else {
+    std::tie(PiKernel, std::ignore, EliminatedArgMask, std::ignore) =
+        sycl::detail::ProgramManager::getInstance().getOrCreateKernel(
+            ContextImpl, DeviceImpl, ExecCG.MKernelName);
+  }
+
+  // Remove eliminated args
+  std::vector<sycl::detail::ArgDesc> MaskedArgs;
+  MaskedArgs.reserve(NodeArgs.size());
+
+  sycl::detail::applyFuncOnFilteredArgs(
+      EliminatedArgMask, NodeArgs,
+      [&MaskedArgs](sycl::detail::ArgDesc &Arg, int NextTrueIndex) {
+        MaskedArgs.emplace_back(Arg.MType, Arg.MPtr, Arg.MSize, NextTrueIndex);
+      });
+
+  // Remember this information before the range dimensions are reversed
+  const bool HasLocalSize = (NDRDesc.LocalSize[0] != 0);
+
+  // Reverse kernel dims
+  sycl::detail::ReverseRangeDimensionsForKernel(NDRDesc);
+
+  size_t RequiredWGSize[3] = {0, 0, 0};
+  size_t *LocalSize = nullptr;
+
+  if (HasLocalSize)
+    LocalSize = &NDRDesc.LocalSize[0];
+  else {
```

Comment:
I think we can remove the bool from before the reversal but at least some (possibly all) backends expect `nullptr` for LocalSize if it is not set by the user, so that cannot be changed here.

Reply:
Fair point! 👍